call-home/src/bin/callhome/collector/report_models.rs

```diff
@@ -184,3 +199,164 @@ fn get_pools_size_vector(pools: Vec<openapi::models::Pool>) -> Vec<u64> {
+                .context(errors::PrometheusOutPutParseFailure)?;
+            Ok(EventsRecord::new(metrics.samples))
+        }
+        Err(_) => errors::StatsFetchFailure.fail(),
+    }
+}
+
+async fn fetch_stats_with_timeout(
+    url: Url,
+) -> Result<reqwest::Response, reqwest_middleware::Error> {
+    new_client()
+        .get(url)
+        .timeout(std::time::Duration::from_secs(30))
+        .send()
+        .await
+}
+
+fn new_client() -> ClientWithMiddleware {
+    // Retry up to 20 times with increasing intervals between attempts.
+    let retry_policy = ExponentialBackoff::builder().build_with_max_retries(20);
+    let client_config = reqwest::Client::new();
+    ClientBuilder::new(client_config)
+        .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+        .build()
+}
+
+struct Record<'a>(&'a Sample);
+
+impl Record<'_> {
+    fn action(&self) -> Action {
+        Action::from(self.0.labels.get(ACTION).unwrap_or(""))
+    }
+    fn resource(&self) -> Resource {
+        Resource::from(self.0.metric.as_str())
+    }
+    fn counter_value(&self) -> u32 {
+        match self.0.value {
+            Value::Counter(v) => v as u32,
+            _ => 0,
+        }
+    }
+    fn try_counter(&self, resource: Resource, action: Action) -> Result<CounterValue, ()> {
+        if self.resource() == resource && self.action() == action {
+            return Ok(CounterValue(self.counter_value()));
+        }
+        Err(())
+    }
+}
+
+/// Metrics Resource.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+pub enum Resource {
+    #[default]
+    Unknown,
+    Pool,
+    Volume,
+}
+/// Metrics action.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+pub enum Action {
+    #[default]
+    Unknown,
+    Created,
+    Deleted,
+}
+
+impl From<&str> for Resource {
+    fn from(s: &str) -> Self {
+        match s {
+            "pool" => Resource::Pool,
+            "volume" => Resource::Volume,
+            _ => Resource::Unknown,
+        }
+    }
+}
+impl From<&str> for Action {
+    fn from(s: &str) -> Self {
+        match s {
+            "created" => Action::Created,
+            "deleted" => Action::Deleted,
+            _ => Action::Unknown,
+        }
+    }
+}
+
+macro_rules! make_counter {
+    ($name:ident, $resource:expr, $action:expr) => {
+        #[derive(Debug, Default, Clone)]
+        pub struct $name(CounterValue);
+        impl std::ops::Deref for $name {
+            type Target = CounterValue;
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+        impl TryFrom<Record<'_>> for $name {
+            type Error = ();
+            fn try_from(record: Record<'_>) -> Result<Self, ()> {
+                record.try_counter($resource, $action).map($name)
+            }
+        }
+    };
+}
+
+/// A Counter represents a single numerical value that only increases.
+#[derive(Debug, Default, Clone)]
+pub struct CounterValue(u32);
```

Comment:
pub(crate) ?

Reply:
Upon making the Counter type as `pub (crate)` it was failing as the `Deref `trait for the counter types inside the make_counter! macro, attempts to use CounterValue as the target type and since CounterValue is pub crate, it cannot be used as the target type in the Deref implementation. Thus leaving it as pub.